typing/typecore.ml

```diff
@@ -5582,13 +5587,30 @@ let report_error ~loc env = function
            fprintf ppf "This expression has type")
         (function ppf ->
            fprintf ppf "but an expression was expected of type");
-  | Apply_non_function typ ->
+  | Apply_non_function (typ, arg_loc, prev_arg_loc) ->
       begin match get_desc typ with
         Tarrow _ ->
-          Location.errorf ~loc
+          let app_loc =
+            (* Span the application, including the extra argument. *)
+            { Location.loc_start = loc.Location.loc_start;
+              loc_end = arg_loc.loc_end;
+              loc_ghost = false }
+          and tail_loc =
+            (* Possible location for a ';'. *)
+            let loc_start = prev_arg_loc.loc_end in
+            { Location.loc_start;
+              loc_end = { loc_start with pos_cnum = loc_start.pos_cnum + 1 };
+              loc_ghost = false }
+          in
+          let sub = [
+            Location.msg ~loc:tail_loc "Maybe you forgot a ';'.";
+            Location.msg ~loc:arg_loc "This argument is not expected.";
```

Comment:
Maybe we could be even more specific: 
```
The function %s cannot be applied to this extra-argument?
```
if we have a identifier at hand and
```
The function cannot be applied to this extra-argument?
```
otherwise?

Reply:
That would indeed be better, now that the location is not pointed to the function. I'll investigate how to get that identifier.

Reply:
I've implemented this and the other suggestions too.
I've handled other cases than `Texp_ident` for fun and I think this function could evolve to a "describe" function if needed for other error messages. (eg. to returns things like `function f`, `method m`, `module M`)